// @pact-foundation/pact-node/src/pact-standalone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var cwd = path.resolve(__dirname, "..");
exports.PACT_STANDALONE_VERSION = "1.33.0";
function getPlatformFolderName(platform, arch) {
    return "" + platform + (platform === "linux" ? "-" + arch : "") + "-" + exports.PACT_STANDALONE_VERSION;
}
exports.getPlatformFolderName = getPlatformFolderName;
exports.standalone = function (platform, arch) {
    platform = platform || process.platform;
    arch = arch || process.arch;
    var binName = function (name) { return "" + name + (platform === "win32" ? ".bat" : ""); };
    var mock = binName("pact-mock-service");
    var verify = binName("pact-provider-verifier");
    var broker = binName("pact-broker");
    var stub = binName("pact-stub-service");
    var basePath = path.join("standalone", getPlatformFolderName(platform, arch), "bin");
    return {
        cwd: cwd,
        brokerPath: path.join(basePath, broker),
        brokerFullPath: path.resolve(cwd, basePath, broker).trim(),
        mockServicePath: path.join(basePath, mock),
        mockServiceFullPath: path.resolve(cwd, basePath, mock).trim(),
        stubPath: path.join(basePath, stub),
        stubFullPath: path.resolve(cwd, basePath, stub).trim(),
        verifierPath: path.join(basePath, verify),
        verifierFullPath: path.resolve(cwd, basePath, verify).trim()
    };
};
exports.default = exports.standalone();
